reportMetrics call fails with 404 Not Found

**Describe the bug**

After a table is created and data is written to it for the first time, if you try to overwrite it or write more rows into it, the write fails with the following error:

```
java.io.IOException: Failed to report metrics for table catalog_name.schema_name.table_name (HTTP 404): Status: 404
Description: Not Found

        at io.delta.storage.commit.uccommitcoordinator.UCDeltaTokenBasedRestClient.reportMetrics(UCDeltaTokenBasedRestClient.java:563)
        at org.apache.spark.sql.delta.catalog.UCDeltaCatalogClientImpl.reportMetrics(UCDeltaCatalogClientImpl.scala:512)
        at org.apache.spark.sql.delta.hooks.metrics.UpdateMetricsHook.$anonfun$sendMetricsAsync$4(UpdateMetricsHook.scala:106)
        at org.apache.spark.sql.delta.hooks.metrics.UpdateMetricsHook.$anonfun$sendMetricsAsync$4$adapted(UpdateMetricsHook.scala:105)
        at scala.Option.foreach(Option.scala:437)
        at org.apache.spark.sql.delta.hooks.metrics.UpdateMetricsHook.$anonfun$sendMetricsAsync$3(UpdateMetricsHook.scala:105)
        at scala.runtime.java8.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.scala:18)
        at scala.concurrent.Future$.$anonfun$apply$1(Future.scala:691)
        at scala.concurrent.impl.Promise$Transformation.run(Promise.scala:500)
        at org.apache.spark.sql.delta.util.threads.SparkThreadLocalCapturingRunnable.$anonfun$run$1(SparkThreadLocalForwardingThreadPoolExecutor.scala:119)
        at scala.runtime.java8.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.scala:18)
        at org.apache.spark.sql.delta.util.threads.SparkThreadLocalCapturingHelper.runWithCaptured(SparkThreadLocalForwardingThreadPoolExecutor.scala:77)
        at org.apache.spark.sql.delta.util.threads.SparkThreadLocalCapturingHelper.runWithCaptured$(SparkThreadLocalForwardingThreadPoolExecutor.scala:60)
        at org.apache.spark.sql.delta.util.threads.SparkThreadLocalCapturingRunnable.runWithCaptured(SparkThreadLocalForwardingThreadPoolExecutor.scala:116)
        at org.apache.spark.sql.delta.util.threads.SparkThreadLocalCapturingRunnable.run(SparkThreadLocalForwardingThreadPoolExecutor.scala:119)
        at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1144)
        at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:642)
        at java.base/java.lang.Thread.run(Thread.java:1583)
Caused by: io.unitycatalog.client.ApiException: reportMetrics call failed with: 404 - Status: 404
Description: Not Found

        at io.unitycatalog.client.delta.api.DeltaTablesApi.getApiException(DeltaTablesApi.java:81)
        at io.unitycatalog.client.delta.api.DeltaTablesApi.reportMetricsWithHttpInfo(DeltaTablesApi.java:588)
        at io.unitycatalog.client.delta.api.DeltaTablesApi.reportMetrics(DeltaTablesApi.java:564)
        at io.delta.storage.commit.uccommitcoordinator.UCDeltaTokenBasedRestClient.reportMetrics(UCDeltaTokenBasedRestClient.java:560)
        ... 17 more
```

**To Reproduce**

Steps to reproduce the behavior:

1. Initialize a PySpark Session with the following code:
```
SparkSession.builder
            .appName(app_name)
            .master("local[4]")
            .config("spark.driver.memory", "4g")
            .config("spark.driver.cores", "2")
            .config("spark.executor.memory", "4g")
            .config("spark.python.worker.memory", "4g")
            .config("spark.driver.extraJavaOptions", "-Djava.security.manager=allow")
            .config("spark.executor.extraJavaOptions", "-Djava.security.manager=allow")
            .config('spark.jars.packages', 'io.delta:delta-spark_4.1_2.13:4.3.0,io.unitycatalog:unitycatalog-spark_4.1_2.13:0.5.0')
            .config("spark.sql.extensions", "io.delta.sql.DeltaSparkSessionExtension")
            .config("spark.sql.catalog.spark_catalog", "org.apache.spark.sql.delta.catalog.DeltaCatalog") 
            .config(f"spark.sql.catalog.catalog_name", "io.unitycatalog.spark.UCSingleCatalog")
            .config(f"spark.sql.catalog.catalog_name.uri", "http://localhost:8080")
            .config(f"spark.sql.catalog.catalog_name.token", "")
            .config("spark.sql.defaultCatalog", "catalog_name")
            .config("spark.sql.defaultSchema", "schema_name")
            .getOrCreate()
```

2. Write a data frame twice: `dataframe.write.format("delta").mode(mode).option("overwriteSchema", "true").saveAsTable("table_name")`

**Expected behavior**

The table is written without errors

**System [please complete the following information]:**

-  Unity Catalog 0.5.0
- pyspark 4.1.2
- Delta Lake 4.3.0

--- a/server/src/main/java/io/unitycatalog/server/service/delta/DeltaApiService.java
+++ b/server/src/main/java/io/unitycatalog/server/service/delta/DeltaApiService.java
@@ -26,6 +26,7 @@ import io.unitycatalog.server.delta.model.DeltaCreateTableRequest;
 import io.unitycatalog.server.delta.model.DeltaCredentialOperation;
 import io.unitycatalog.server.delta.model.DeltaCredentialsResponse;
 import io.unitycatalog.server.delta.model.DeltaLoadTableResponse;
+import io.unitycatalog.server.delta.model.DeltaReportMetricsRequest;
 import io.unitycatalog.server.delta.model.DeltaStagingTableResponse;
 import io.unitycatalog.server.delta.model.DeltaTableType;
 import io.unitycatalog.server.delta.model.DeltaUpdateTableRequest;
@@ -263,6 +264,42 @@ public class DeltaApiService extends AuthorizedService {
       @Param("table") @AuthorizeResourceKey(TABLE) String table,
       DeltaUpdateTableRequest request) {
     return tableRepository.updateTableForDelta(catalog, schema, table, request);
+  }
+
+  // ==================== Report Metrics API ====================
+
+  /**
+   * Accept commit metrics (telemetry) reported by Delta clients after a commit (e.g. Delta Spark's
+   * UpdateMetricsHook). Per {@code delta.yaml}, the body's {@code table-id} must identify the same
+   * table as the path; mismatches are rejected. The metrics themselves are currently acknowledged
+   * and discarded -- the endpoint exists so post-commit metrics reporting succeeds instead of
+   * failing writes with 404.
+   */
+  @Post("/delta/v1/catalogs/{catalog}/schemas/{schema}/tables/{table}/metrics")
+  @AuthorizeExpression(AuthorizeExpressions.UPDATE_TABLE)
+  public HttpResponse reportMetrics(
+      @Param("catalog") @AuthorizeResourceKey(CATALOG) String catalog,
+      @Param("schema") @AuthorizeResourceKey(SCHEMA) String schema,
+      @Param("table") @AuthorizeResourceKey(TABLE) String table,
+      DeltaReportMetricsRequest request) {
+    if (request == null || request.getTableId() == null) {
+      throw new BaseException(
+          ErrorCode.INVALID_ARGUMENT, "Must supply table-id in the request body.");
+    }
+    TableInfoDAO tableInfo = tableRepository.findTableOrThrow(catalog, schema, table);
+    if (!tableInfo.getId().equals(request.getTableId())) {
+      throw new BaseException(
+          ErrorCode.INVALID_ARGUMENT,
+          "table-id "
+              + request.getTableId()
+              + " does not match table "
+              + catalog
+              + "."
+              + schema
+              + "."
+              + table);
+    }
+    return HttpResponse.of(HttpStatus.NO_CONTENT);
   }
 
   // ==================== Table Credentials API ====================

--- a/server/src/main/java/io/unitycatalog/server/service/delta/DeltaApiService.java
+++ b/server/src/main/java/io/unitycatalog/server/service/delta/DeltaApiService.java
@@ -27,6 +27,7 @@ import io.unitycatalog.server.delta.model.DeltaCredentialOperation;
 import io.unitycatalog.server.delta.model.DeltaCredentialsResponse;
 import io.unitycatalog.server.delta.model.DeltaLoadTableResponse;
 import io.unitycatalog.server.delta.model.DeltaRenameTableRequest;
+import io.unitycatalog.server.delta.model.DeltaReportMetricsRequest;
 import io.unitycatalog.server.delta.model.DeltaStagingTableResponse;
 import io.unitycatalog.server.delta.model.DeltaTableType;
 import io.unitycatalog.server.delta.model.DeltaUpdateTableRequest;
@@ -284,6 +285,36 @@ public class DeltaApiService extends AuthorizedService {
       throw new BaseException(ErrorCode.INVALID_ARGUMENT, "New table name is required.");
     }
     tableRepository.renameTable(catalog, schema, table, request.getNewName());
+    return HttpResponse.of(HttpStatus.NO_CONTENT);
+  }
+
+  // ==================== Report Metrics API ====================
+
+  /**
+   * Accept commit metrics reported by Delta clients after a commit (e.g. Delta Spark's {@code
+   * UpdateMetricsHook}). Per {@code delta.yaml} the body's {@code table-id} must identify the same
+   * table as the path; mismatches are rejected. The metrics are acknowledged and discarded -- UC
+   * does not persist client telemetry today. Authorization mirrors {@link #updateTable}, so a
+   * caller who can commit to the table can report metrics for it.
+   */
+  @Post("/delta/v1/catalogs/{catalog}/schemas/{schema}/tables/{table}/metrics")
+  @AuthorizeExpression(AuthorizeExpressions.UPDATE_TABLE)
+  public HttpResponse reportMetrics(
+      @Param("catalog") @AuthorizeResourceKey(CATALOG) String catalog,
+      @Param("schema") @AuthorizeResourceKey(SCHEMA) String schema,
+      @Param("table") @AuthorizeResourceKey(TABLE) String table,
+      DeltaReportMetricsRequest request) {
+    if (request == null || request.getTableId() == null) {
+      throw new BaseException(ErrorCode.INVALID_ARGUMENT, "table-id is required.");
+    }
+    TableInfoDAO tableInfo = tableRepository.findTableOrThrow(catalog, schema, table);
+    if (!tableInfo.getId().equals(request.getTableId())) {
+      throw new BaseException(
+          ErrorCode.INVALID_ARGUMENT,
+          String.format(
+              "table-id %s does not match table %s.%s.%s.",
+              request.getTableId(), catalog, schema, table));
+    }
     return HttpResponse.of(HttpStatus.NO_CONTENT);
   }
 

--- a/server/src/test/java/io/unitycatalog/server/sdk/delta/SdkReportMetricsTest.java
+++ b/server/src/test/java/io/unitycatalog/server/sdk/delta/SdkReportMetricsTest.java
@@ -1,0 +1,146 @@
+package io.unitycatalog.server.sdk.delta;
+
+import static io.unitycatalog.server.utils.TestUtils.assertApiExceptionStatusOnly;
+import static org.assertj.core.api.Assertions.assertThat;
+
+import io.unitycatalog.client.ApiException;
+import io.unitycatalog.client.ApiResponse;
+import io.unitycatalog.client.delta.api.DeltaTablesApi;
+import io.unitycatalog.client.delta.model.DeltaErrorType;
+import io.unitycatalog.client.delta.model.DeltaReportMetricsRequest;
+import io.unitycatalog.client.model.ColumnInfo;
+import io.unitycatalog.client.model.ColumnTypeName;
+import io.unitycatalog.client.model.CreateCatalog;
+import io.unitycatalog.client.model.CreateSchema;
+import io.unitycatalog.client.model.CreateTable;
+import io.unitycatalog.client.model.DataSourceFormat;
+import io.unitycatalog.client.model.TableInfo;
+import io.unitycatalog.client.model.TableType;
+import io.unitycatalog.server.base.BaseServerTest;
+import io.unitycatalog.server.base.catalog.CatalogOperations;
+import io.unitycatalog.server.base.schema.SchemaOperations;
+import io.unitycatalog.server.base.table.TableOperations;
+import io.unitycatalog.server.sdk.catalog.SdkCatalogOperations;
+import io.unitycatalog.server.sdk.schema.SdkSchemaOperations;
+import io.unitycatalog.server.sdk.tables.SdkTableOperations;
+import io.unitycatalog.server.utils.TestUtils;
+import java.io.IOException;
+import java.util.List;
+import java.util.Optional;
+import java.util.UUID;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
+
+/**
+ * End-to-end tests for the UC Delta API reportMetrics endpoint. Regression test for the missing
+ * endpoint (issue #1654): the server advertised {@code POST .../tables/{table}/metrics} in its
+ * config endpoint list but did not implement it, so every post-commit metrics report from Delta
+ * clients failed with 404 and surfaced as a failed write.
+ */
+public class SdkReportMetricsTest extends BaseServerTest {
+
+  private static final String TABLE_NAME = "metrics_test_table";
+
+  private CatalogOperations catalogOps;
+  private SchemaOperations schemaOps;
+  private TableOperations tableOps;
+  private DeltaTablesApi deltaTablesApi;
+
+  @BeforeEach
+  public void setUp() {
+    super.setUp();
+    var apiClient = TestUtils.createApiClient(serverConfig);
+    catalogOps = new SdkCatalogOperations(apiClient);
+    schemaOps = new SdkSchemaOperations(apiClient);
+    tableOps = new SdkTableOperations(apiClient);
+    deltaTablesApi = new DeltaTablesApi(apiClient);
+    cleanUp();
+    createCatalogAndSchema();
+  }
+
+  private void cleanUp() {
+    try {
+      catalogOps.deleteCatalog(TestUtils.CATALOG_NAME, Optional.of(true));
+    } catch (Exception e) {
+      // Ignore
+    }
+  }
+
+  private void createCatalogAndSchema() {
+    try {
+      catalogOps.createCatalog(new CreateCatalog().name(TestUtils.CATALOG_NAME).comment("test"));
+      schemaOps.createSchema(
+          new CreateSchema().name(TestUtils.SCHEMA_NAME).catalogName(TestUtils.CATALOG_NAME));
+    } catch (ApiException e) {
+      throw new RuntimeException(e);
+    }
+  }
+
+  private UUID createTable() throws ApiException, IOException {
+    TableInfo table =
+        tableOps.createTable(
+            new CreateTable()
+                .name(TABLE_NAME)
+                .catalogName(TestUtils.CATALOG_NAME)
+                .schemaName(TestUtils.SCHEMA_NAME)
+                .tableType(TableType.EXTERNAL)
+                .dataSourceFormat(DataSourceFormat.DELTA)
+                .storageLocation("file:///tmp/uc_test/" + TABLE_NAME)
+                .columns(
+                    List.of(
+                        new ColumnInfo()
+                            .name("id")
+                            .typeName(ColumnTypeName.LONG)
+                            .typeText("bigint")
+                            .typeJson(
+                                "{\"name\":\"id\",\"type\":\"long\",\"nullable\":true,\"metadata\":{}}")
+                            .position(0))));
+    return UUID.fromString(table.getTableId());
+  }
+
+  @Test
+  public void testReportMetrics() throws Exception {
+    UUID tableId = createTable();
+
+    // Success: matching table-id is acknowledged with 204 No Content
+    ApiResponse<Void> response =
+        deltaTablesApi.reportMetricsWithHttpInfo(
+            TestUtils.CATALOG_NAME,
+            TestUtils.SCHEMA_NAME,
+            TABLE_NAME,
+            new DeltaReportMetricsRequest().tableId(tableId));
+    assertThat(response.getStatusCode()).isEqualTo(204);
+
+    // Error: table-id not matching the path table returns 400
+    TestUtils.assertDeltaApiException(
+        () ->
+            deltaTablesApi.reportMetrics(
+                TestUtils.CATALOG_NAME,
+                TestUtils.SCHEMA_NAME,
+                TABLE_NAME,
+                new DeltaReportMetricsRequest().tableId(UUID.randomUUID())),
+        DeltaErrorType.INVALID_PARAMETER_VALUE_EXCEPTION,
+        "does not match");
+
+    // Error: missing table-id returns 400
+    TestUtils.assertDeltaApiException(
+        () ->
+            deltaTablesApi.reportMetrics(
+                TestUtils.CATALOG_NAME,
+                TestUtils.SCHEMA_NAME,
+                TABLE_NAME,
+                new DeltaReportMetricsRequest()),
+        DeltaErrorType.INVALID_PARAMETER_VALUE_EXCEPTION,
+        "table-id");
+
+    // Error: nonexistent table returns 404
+    assertApiExceptionStatusOnly(
+        () ->
+            deltaTablesApi.reportMetrics(
+                TestUtils.CATALOG_NAME,
+                TestUtils.SCHEMA_NAME,
+                "nonexistent_table",
+                new DeltaReportMetricsRequest().tableId(tableId)),
+        404);
+  }
+}

--- a/server/src/test/java/io/unitycatalog/server/sdk/delta/SdkReportMetricsTest.java
+++ b/server/src/test/java/io/unitycatalog/server/sdk/delta/SdkReportMetricsTest.java
@@ -1,0 +1,82 @@
+package io.unitycatalog.server.sdk.delta;
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+import io.unitycatalog.client.ApiResponse;
+import io.unitycatalog.client.delta.model.DeltaErrorType;
+import io.unitycatalog.client.delta.model.DeltaReportMetricsRequest;
+import io.unitycatalog.server.base.ServerConfig;
+import io.unitycatalog.server.base.catalog.CatalogOperations;
+import io.unitycatalog.server.base.delta.DeltaBaseTableCRUDTestEnv;
+import io.unitycatalog.server.base.schema.SchemaOperations;
+import io.unitycatalog.server.sdk.catalog.SdkCatalogOperations;
+import io.unitycatalog.server.sdk.schema.SdkSchemaOperations;
+import io.unitycatalog.server.utils.TestUtils;
+import java.util.Map;
+import java.util.UUID;
+import org.junit.jupiter.api.Test;
+
+/**
+ * E2E tests for the Delta REST Catalog {@code POST .../metrics} endpoint (204 on success, 400 when
+ * the body {@code table-id} is missing or does not match the path table, 404 when the table is
+ * missing).
+ */
+public class SdkReportMetricsTest extends DeltaBaseTableCRUDTestEnv {
+
+  @Override
+  protected CatalogOperations createCatalogOperations(ServerConfig serverConfig) {
+    return new SdkCatalogOperations(TestUtils.createApiClient(serverConfig));
+  }
+
+  @Override
+  protected SchemaOperations createSchemaOperations(ServerConfig serverConfig) {
+    return new SdkSchemaOperations(TestUtils.createApiClient(serverConfig));
+  }
+
+  @Test
+  public void testReportMetricsEndpoint() throws Exception {
+    Handle table = createDeltaManaged("metricsTable", Map.of());
+
+    // A matching table-id is acknowledged with 204 No Content.
+    ApiResponse<Void> response =
+        deltaTablesApi.reportMetricsWithHttpInfo(
+            TestUtils.CATALOG_NAME,
+            TestUtils.SCHEMA_NAME,
+            table.name(),
+            new DeltaReportMetricsRequest().tableId(table.tableId()));
+    assertThat(response.getStatusCode()).isEqualTo(204);
+
+    // A missing table-id is rejected before the table is looked up.
+    TestUtils.assertDeltaApiException(
+        () ->
+            deltaTablesApi.reportMetrics(
+                TestUtils.CATALOG_NAME,
+                TestUtils.SCHEMA_NAME,
+                table.name(),
+                new DeltaReportMetricsRequest()),
+        DeltaErrorType.INVALID_PARAMETER_VALUE_EXCEPTION,
+        "table-id is required");
+
+    // A table-id that resolves to a different table than the path is rejected.
+    TestUtils.assertDeltaApiException(
+        () ->
+            deltaTablesApi.reportMetrics(
+                TestUtils.CATALOG_NAME,
+                TestUtils.SCHEMA_NAME,
+                table.name(),
+                new DeltaReportMetricsRequest().tableId(UUID.randomUUID())),
+        DeltaErrorType.INVALID_PARAMETER_VALUE_EXCEPTION,
+        "does not match table");
+
+    // A missing table returns 404 even when the table-id is well-formed.
+    TestUtils.assertDeltaApiException(
+        () ->
+            deltaTablesApi.reportMetrics(
+                TestUtils.CATALOG_NAME,
+                TestUtils.SCHEMA_NAME,
+                "missingTable",
+                new DeltaReportMetricsRequest().tableId(table.tableId())),
+        DeltaErrorType.NO_SUCH_TABLE_EXCEPTION,
+        "Table not found");
+  }
+}

